core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyRedundantCase.java

```diff
@@ -47,20 +58,28 @@ public Optional<Expression> apply(Expression expression, Session session, Map<Sy
             return Optional.empty();
         }
 
-        // TODO: generalize to arbitrary number of clauses
-        if (caseTerm.whenClauses().size() != 1) {
+        Expression defaultValue = caseTerm.defaultValue();
+        if (!caseTerm.whenClauses().stream().map(WhenClause::getResult).allMatch(result -> result.equals(TRUE) || result.equals(FALSE)) ||
+                (!defaultValue.equals(TRUE) && !defaultValue.equals(FALSE)) ||
+                caseTerm.whenClauses().stream().map(WhenClause::getOperand).anyMatch(e -> !isDeterministic(e))) {
```

Comment:
Would it be efficient if we could have the deterministic check first before apply the checks on the default value and its result ?

Reply:
Probably not. Checking whether an expression is deterministic is more expensive, as it needs to analyze it recursively.